velox/common/memory/MallocAllocator.cpp

```diff
@@ -65,30 +65,43 @@ bool MallocAllocator::allocateNonContiguousWithoutRetry(
 
   std::vector<void*> pages;
   pages.reserve(mix.numSizes);
+  bool failedToAllocate = false;
   for (int32_t i = 0; i < mix.numSizes; ++i) {
     // Trigger allocation failure by breaking out the loop.
     if (testingHasInjectedFailure(InjectedFailure::kAllocate)) {
+      failedToAllocate = true;
       break;
     }
-    MachinePageCount numSizeClassPages =
-        mix.sizeCounts[i] * sizeClassSizes_[mix.sizeIndices[i]];
-    void* ptr;
-    stats_.recordAllocate(
-        AllocationTraits::pageBytes(sizeClassSizes_[mix.sizeIndices[i]]),
-        mix.sizeCounts[i],
-        [&]() {
-          ptr = ::malloc(
-              AllocationTraits::pageBytes(numSizeClassPages)); // NOLINT
-        });
-    if (ptr == nullptr) {
-      // Failed to allocate memory from memory.
-      break;
+
+    const auto maxClassPagesInRun = static_cast<int>(
+        Allocation::PageRun::kMaxPagesInRun /
+        sizeClassSizes_[mix.sizeIndices[i]]);
+    auto totalClassPages = mix.sizeCounts[i];
+    while (totalClassPages > 0) {
+      const auto numClassPagesToAllocate =
+          std::min(totalClassPages, maxClassPagesInRun);
+      const MachinePageCount numSizeClassPages =
+          numClassPagesToAllocate * sizeClassSizes_[mix.sizeIndices[i]];
+      totalClassPages -= maxClassPagesInRun;
+      void* ptr;
+      stats_.recordAllocate(
+          AllocationTraits::pageBytes(sizeClassSizes_[mix.sizeIndices[i]]),
+          numClassPagesToAllocate,
+          [&]() {
+            ptr = ::malloc(
+                AllocationTraits::pageBytes(numSizeClassPages)); // NOLINT
+          });
+      if (ptr == nullptr) {
+        // Failed to allocate memory from memory.
+        failedToAllocate = true;
+        break;
+      }
+      pages.emplace_back(ptr);
+      out.append(reinterpret_cast<uint8_t*>(ptr), numSizeClassPages); // NOLINT
     }
-    pages.emplace_back(ptr);
-    out.append(reinterpret_cast<uint8_t*>(ptr), numSizeClassPages); // NOLINT
   }
 
-  if (pages.size() != mix.numSizes) {
```

Comment:
I think we can keep this check instead of defining a new flag 'failedToAllocate'? Thanks!

Reply:
The number of pages allocated for a given size can be more than 1. So it will not be equal to min.numSizes.

Reply:
I see. Then you may check 
```
if (out.numPages() != mix.totalPages) {
}
```